src/System.Private.CoreLib/src/System/Threading/ThreadPool.Unix.cs

```diff
@@ -4,18 +4,206 @@
     {
-        public bool Unregister(WaitHandle waitObject)
+        internal RegisteredWaitHandle(WaitHandle waitHandle, _ThreadPoolWaitOrTimerCallback callbackHelper,
+            int millisecondsTimeout, bool repeating)
         {
-            // UNIXTODO: ThreadPool
-            throw new NotImplementedException();
+            Handle = waitHandle;
+            Callback = callbackHelper;
+            TimeoutDurationMs = millisecondsTimeout;
+            Repeating = repeating;
+            RestartTimeout(Environment.TickCount);
+        }
+
+        /// <summary>
+        /// The callback to execute when the wait on <see cref="Handle"/> either times out or completes.
+        /// </summary>
+        internal _ThreadPoolWaitOrTimerCallback Callback { get; }
+
+
+        /// <summary>
+        /// The <see cref="WaitHandle"/> that was registered.
+        /// </summary>
+        internal WaitHandle Handle { get; }
+
+        /// <summary>
+        /// The time this handle times out at in ms.
+        /// </summary>
+        internal int TimeoutTimeMs { get; private set; }
+
+        private int TimeoutDurationMs { get; }
+
+        internal bool InfiniteTimeout => TimeoutDurationMs == -1;
+
+        internal void RestartTimeout(int currentTimeMs)
+        {
+            TimeoutTimeMs = currentTimeMs + TimeoutDurationMs;
+        }
+
+        /// <summary>
+        /// Whether or not the wait is a repeating wait.
+        /// </summary>
+        internal bool Repeating { get; }
+
+        /// <summary>
+        /// The <see cref="WaitHandle"/> the user passed in via <see cref="Unregister(WaitHandle)"/>.
+        /// </summary>
+        private SafeWaitHandle UserUnregisterWaitHandle { get; set; } = new SafeWaitHandle((IntPtr)(-1), false); // Initialize with an invalid handle like CoreCLR
+
+        private IntPtr UserUnregisterWaitHandleValue { get; set; } = new IntPtr(-1);
+
+        /// <summary>
+        /// Whether or not <see cref="UserUnregisterWaitHandle"/> has been signaled yet.
+        /// </summary>
+        private volatile int _unregisterSignaled;
+
+        internal bool IsBlocking { get; set; } = true;
+
+        /// <summary>
+        /// The <see cref="ClrThreadPool.WaitThread"/> this <see cref="RegisteredWaitHandle"/> was registered on.
+        /// </summary>
+        internal ClrThreadPool.WaitThread WaitThread { get; set; }
+
+        private volatile int _numRequestedCallbacks;
+
+        private LowLevelLock _callbackLock = new LowLevelLock();
+
+        private bool _signalAfterCallbacksComplete;
+
+        private int _unregisterCalled;
+
+        private AutoResetEvent _unregisteredEvent = new AutoResetEvent(false);
+
+        internal bool Unregister(WaitHandle waitObject)
+        {
+            if (Interlocked.Exchange(ref _unregisterCalled, 1) == 0)
+            {
+                UserUnregisterWaitHandle = waitObject?.SafeWaitHandle;
+                if (!(UserUnregisterWaitHandle?.IsInvalid ?? true))
```

Comment:
In my opinion, (UserUnregisterWaitHandle != null && !UserUnregisterWaitHandle.IsInvalid) is easier to read and understand, but up to you